{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chapter 3: Biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (repodata.json): ...working... done\n",
      "Solving environment: ...working... done\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n",
      "\n",
      "Intel(R) MPI Library 2019 Update 6 for Windows* Target Build Environment for Intel(R) 64 applications\n",
      "Copyright 2007-2019 Intel Corporation.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!conda install biopython "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.72\n"
     ]
    }
   ],
   "source": [
    "import Bio\n",
    "print(Bio.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Biopython seq object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Bio.Seq.Seq'>\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "\n",
    "my_seq = Seq(\"ATGGCCTTAAA\")\n",
    "print(type(my_seq))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alphabet()\n",
      "DNAAlphabet()\n",
      "RNAAlphabet()\n",
      "ProteinAlphabet()\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "from Bio.Alphabet import generic_dna, generic_protein, generic_rna\n",
    "my_seq = Seq(\"ATGGCCTTAAA\")\n",
    "\n",
    "\n",
    "dna = Seq(\"ATGGCCTTAAA\", generic_dna)\n",
    "rna = Seq(\"ACCCGUUGU\", generic_rna)\n",
    "protein = Seq(\"AKKKGGGUUULL\", generic_protein)\n",
    "print(my_seq.alphabet)\n",
    "print(dna.alphabet)\n",
    "print(rna.alphabet)\n",
    "print(protein.alphabet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AUGGCCUUAAAU\n",
      "RNAAlphabet()\n",
      "MALN\n",
      "ExtendedIUPACProtein()\n"
     ]
    }
   ],
   "source": [
    "# Central dogma example.\n",
    "gene = Seq(\"ATGGCCTTAAAT\", generic_dna)\n",
    "\n",
    "#transcription\n",
    "\n",
    "transcript = gene.transcribe()\n",
    "print(transcript)\n",
    "print(transcript.alphabet)\n",
    "\n",
    "\n",
    "#Translarion\n",
    "protein = transcript.translate()\n",
    "print(protein)\n",
    "print(protein.alphabet)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Biopython SeqRecords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: P62945\n",
      "Name: RL41_HUMAN\n",
      "Description: 60S ribosomal protein L41\n",
      "Number of features: 0\n",
      "Seq('MRAKWRKKRMRRLKRKRRKMRQRSK', ProteinAlphabet())\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import generic_protein\n",
    "protein_record = SeqRecord(Seq(\"MRAKWRKKRMRRLKRKRRKMRQRSK\",\n",
    "                       generic_protein),\n",
    "                   id=\"P62945\", name=\"RL41_HUMAN\",\n",
    "                   description=\"60S ribosomal protein L41\")\n",
    "print(protein_record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parsing sequence files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: >sp|Q9SE35|20-107\n",
      "Name: >sp|Q9SE35|20-107\n",
      "Description: >sp|Q9SE35|20-107\n",
      "Number of features: 0\n",
      "Seq('QSIADLAAANLSTEDSKSAQLISADSSDDASDSSVESVDAASSDVSGSSVESVD...RIL', ProteinAlphabet())\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import generic_protein\n",
    "sequence = ''\n",
    "with open('sample.fasta','r') as f:\n",
    "    lines = f.readlines()\n",
    "    \n",
    "    for line in lines:\n",
    "        if line.startswith('>'):\n",
    "            description = line.rstrip()\n",
    "        else:\n",
    "            sequence = sequence + line.rstrip()\n",
    "\n",
    "protein_record = SeqRecord(Seq(sequence,generic_protein),\n",
    "                   id=description , name=description ,\n",
    "                   description=description)\n",
    "print(protein_record)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### With SeqIO module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: sp|Q9SE35|20-107\n",
      "Name: sp|Q9SE35|20-107\n",
      "Description: sp|Q9SE35|20-107\n",
      "Number of features: 0\n",
      "Seq('QSIADLAAANLSTEDSKSAQLISADSSDDASDSSVESVDAASSDVSGSSVESVD...RIL', SingleLetterAlphabet())\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "records = SeqIO.parse(\"sample.fasta\", \"fasta\")\n",
    "for record in records:\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fasta file with multiple sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sp|Q3ZM63|ETDA_HUMAN\n",
      "sp|P53803|RPAB4_HUMAN\n",
      "sp|Q538Z0|LUZP6_HUMAN\n",
      "sp|Q9BZ97|TTY13_HUMAN\n",
      "sp|P58511|SI11A_HUMAN\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "record_list = []\n",
    "for record in SeqIO.parse(\"multi_sequence.fasta\", \"fasta\"):\n",
    "    print(record.id)\n",
    "    record_list.append(record)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GenBank File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: X03109.1\n",
      "Name: X03109\n",
      "Description: Chimpanzee fetal G-gamma-globin gene\n",
      "Number of features: 13\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=14-NOV-2006\n",
      "/accessions=['X03109']\n",
      "/sequence_version=1\n",
      "/gi=38219\n",
      "/keywords=['direct repeat', 'G-gamma-globin', 'gamma-globin', 'tandem repeat']\n",
      "/source=Pan troglodytes (chimpanzee)\n",
      "/organism=Pan troglodytes\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Pan']\n",
      "/references=[Reference(title='Chimpanzee fetal G gamma and A gamma globin gene nucleotide sequences provide further evidence of gene conversions in hominine evolution', ...), Reference(title='Direct Submission', ...)]\n",
      "Seq('CCGGCGGCTGGCTAGGGATGAAGAATAAAAGGAAGCACCCTCCAGCAGTTCCAC...AAT', IUPACAmbiguousDNA())\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from Bio import SeqIO\n",
    "records = SeqIO.parse(\"sequence.gb\",\"genbank\")\n",
    "for record in records:\n",
    "    print(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing Sequence files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">P62945 60S ribosomal protein L41\n",
      "MRAKWRKKRMRRLKRKRRKMRQRSK\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import generic_protein\n",
    "protein_record = SeqRecord(Seq(\"MRAKWRKKRMRRLKRKRRKMRQRSK\",\n",
    "                       generic_protein),\n",
    "                   id=\"P62945\", name=\"RL41_HUMAN\",\n",
    "                   description=\"60S ribosomal protein L41\")\n",
    "\n",
    "SeqIO.write(protein_record, \"example.fasta\", \"fasta\")\n",
    "\n",
    "with open('example.fasta') as f:\n",
    "    print(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">P62945 60S ribosomal protein L41\n",
      "MRAKWRKKRMRRLKRKRRKMRQRSK\n",
      ">O00631 Sarcolipin\n",
      "MGINTRELFLNFTIVLITVILMWLLVRSYQY\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import generic_protein\n",
    "record_list = []\n",
    "protein_record_1 = SeqRecord(Seq(\"MRAKWRKKRMRRLKRKRRKMRQRSK\",\n",
    "                       generic_protein),\n",
    "                   id=\"P62945\", name=\"RL41_HUMAN\",\n",
    "                   description=\"60S ribosomal protein L41\")\n",
    "\n",
    "protein_record_2 = SeqRecord(Seq(\"MGINTRELFLNFTIVLITVILMWLLVRSYQY\",\n",
    "                       generic_protein),\n",
    "                   id=\"O00631\", name=\"SARCO_HUMAN\",\n",
    "                   description=\"Sarcolipin\")\n",
    "\n",
    "record_list = [protein_record_1,protein_record_2]\n",
    "\n",
    "SeqIO.write(record_list, \"example_2.fasta\", \"fasta\")\n",
    "\n",
    "with open('example_2.fasta') as f:\n",
    "    print(f.read())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence Alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pairwise Sequence alingmnet (Global)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TACGCCC--GC\n",
      "||-||||--||\n",
      "TA-GCCCATGC\n",
      "\n",
      "8.0\n"
     ]
    }
   ],
   "source": [
    "from Bio import Align\n",
    "aligner = Align.PairwiseAligner()\n",
    "seq1 = 'TACGCCCGC'\n",
    "seq2 = 'TAGCCCATGC'\n",
    "alignments = aligner.align(seq1, seq2)\n",
    "for alignment in alignments:\n",
    "    print(alignment)\n",
    "score = aligner.score(seq1, seq2)\n",
    "\n",
    "\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pairwise sequence aligner with parameters\n",
      "  match_score: 1.000000\n",
      "  mismatch_score: 0.000000\n",
      "  target_internal_open_gap_score: 0.000000\n",
      "  target_internal_extend_gap_score: 0.000000\n",
      "  target_left_open_gap_score: 0.000000\n",
      "  target_left_extend_gap_score: 0.000000\n",
      "  target_right_open_gap_score: 0.000000\n",
      "  target_right_extend_gap_score: 0.000000\n",
      "  query_internal_open_gap_score: 0.000000\n",
      "  query_internal_extend_gap_score: 0.000000\n",
      "  query_left_open_gap_score: 0.000000\n",
      "  query_left_extend_gap_score: 0.000000\n",
      "  query_right_open_gap_score: 0.000000\n",
      "  query_right_extend_gap_score: 0.000000\n",
      "  mode: global\n",
      "\n",
      "----------------\n",
      "Needleman-Wunsch\n"
     ]
    }
   ],
   "source": [
    "from Bio import Align\n",
    "aligner = Align.PairwiseAligner()\n",
    "print(aligner)\n",
    "print('----------------')\n",
    "print(aligner.algorithm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TACGCCC--GC\n",
      "||-||||--||\n",
      "TA-GCCCATGC\n",
      "\n",
      "0.5\n"
     ]
    }
   ],
   "source": [
    "from Bio import Align\n",
    "aligner = Align.PairwiseAligner()\n",
    "seq1 = 'TACGCCCGC'\n",
    "seq2 = 'TAGCCCATGC'\n",
    "aligner.match_score = 1.0\n",
    "aligner.mismatch_score = -2.0\n",
    "aligner.gap_score = -2.5\n",
    "alignments = aligner.align(seq1, seq2)\n",
    "for alignment in alignments:\n",
    "    print(alignment)\n",
    "score = aligner.score(seq1, seq2)\n",
    "\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Local alingmnet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TACGCCC--GC\n",
      "||-||||--||\n",
      "TA-GCCCATGC\n",
      "\n",
      "8.0\n"
     ]
    }
   ],
   "source": [
    "from Bio import Align\n",
    "aligner = Align.PairwiseAligner()\n",
    "aligner.mode = 'local'\n",
    "seq1 = 'TACGCCCGC'\n",
    "seq2 = 'TAGCCCATGC'\n",
    "alignments = aligner.align(seq1, seq2)\n",
    "for alignment in alignments:\n",
    "    print(alignment)\n",
    "score = aligner.score(seq1, seq2)\n",
    "\n",
    "\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TACGCCCGC  \n",
      "   ||||    \n",
      " TAGCCCATGC\n",
      "\n",
      "4.0\n"
     ]
    }
   ],
   "source": [
    "from Bio import Align\n",
    "aligner = Align.PairwiseAligner()\n",
    "aligner.mode = 'local'\n",
    "seq1 = 'TACGCCCGC'\n",
    "seq2 = 'TAGCCCATGC'\n",
    "aligner.match_score = 1.0\n",
    "aligner.mismatch_score = -2.0\n",
    "aligner.gap_score = -2.5\n",
    "alignments = aligner.align(seq1, seq2)\n",
    "for alignment in alignments:\n",
    "    print(alignment)\n",
    "score = aligner.score(seq1, seq2)\n",
    "\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['BENNER22',\n",
       " 'BENNER6',\n",
       " 'BENNER74',\n",
       " 'BLOSUM45',\n",
       " 'BLOSUM50',\n",
       " 'BLOSUM62',\n",
       " 'BLOSUM80',\n",
       " 'BLOSUM90',\n",
       " 'DAYHOFF',\n",
       " 'FENG',\n",
       " 'GENETIC',\n",
       " 'GONNET1992',\n",
       " 'JOHNSON',\n",
       " 'JONES',\n",
       " 'LEVIN',\n",
       " 'MCLACHLAN',\n",
       " 'MDM78',\n",
       " 'NUC.4.4',\n",
       " 'PAM250',\n",
       " 'PAM30',\n",
       " 'PAM70',\n",
       " 'RAO',\n",
       " 'RISLER',\n",
       " 'SCHNEIDER',\n",
       " 'STR']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Align import substitution_matrices\n",
    "substitution_matrices.load()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#  Matrix made by matblas from blosum62.iij\n",
      "#  * column uses minimum score\n",
      "#  BLOSUM Clustered Scoring Matrix in 1/2 Bit Units\n",
      "#  Blocks Database = /data/blocks_5.0/blocks.dat\n",
      "#  Cluster Percentage: >= 62\n",
      "#  Entropy =   0.6979, Expected =  -0.5209\n",
      "     A    R    N    D    C    Q    E    G    H    I    L    K    M    F    P    S    T    W    Y    V    B    Z    X    *\n",
      "A  4.0 -1.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  0.0 -3.0 -2.0  0.0 -2.0 -1.0  0.0 -4.0\n",
      "R -1.0  5.0  0.0 -2.0 -3.0  1.0  0.0 -2.0  0.0 -3.0 -2.0  2.0 -1.0 -3.0 -2.0 -1.0 -1.0 -3.0 -2.0 -3.0 -1.0  0.0 -1.0 -4.0\n",
      "N -2.0  0.0  6.0  1.0 -3.0  0.0  0.0  0.0  1.0 -3.0 -3.0  0.0 -2.0 -3.0 -2.0  1.0  0.0 -4.0 -2.0 -3.0  3.0  0.0 -1.0 -4.0\n",
      "D -2.0 -2.0  1.0  6.0 -3.0  0.0  2.0 -1.0 -1.0 -3.0 -4.0 -1.0 -3.0 -3.0 -1.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
      "C  0.0 -3.0 -3.0 -3.0  9.0 -3.0 -4.0 -3.0 -3.0 -1.0 -1.0 -3.0 -1.0 -2.0 -3.0 -1.0 -1.0 -2.0 -2.0 -1.0 -3.0 -3.0 -2.0 -4.0\n",
      "Q -1.0  1.0  0.0  0.0 -3.0  5.0  2.0 -2.0  0.0 -3.0 -2.0  1.0  0.0 -3.0 -1.0  0.0 -1.0 -2.0 -1.0 -2.0  0.0  3.0 -1.0 -4.0\n",
      "E -1.0  0.0  0.0  2.0 -4.0  2.0  5.0 -2.0  0.0 -3.0 -3.0  1.0 -2.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
      "G  0.0 -2.0  0.0 -1.0 -3.0 -2.0 -2.0  6.0 -2.0 -4.0 -4.0 -2.0 -3.0 -3.0 -2.0  0.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -1.0 -4.0\n",
      "H -2.0  0.0  1.0 -1.0 -3.0  0.0  0.0 -2.0  8.0 -3.0 -3.0 -1.0 -2.0 -1.0 -2.0 -1.0 -2.0 -2.0  2.0 -3.0  0.0  0.0 -1.0 -4.0\n",
      "I -1.0 -3.0 -3.0 -3.0 -1.0 -3.0 -3.0 -4.0 -3.0  4.0  2.0 -3.0  1.0  0.0 -3.0 -2.0 -1.0 -3.0 -1.0  3.0 -3.0 -3.0 -1.0 -4.0\n",
      "L -1.0 -2.0 -3.0 -4.0 -1.0 -2.0 -3.0 -4.0 -3.0  2.0  4.0 -2.0  2.0  0.0 -3.0 -2.0 -1.0 -2.0 -1.0  1.0 -4.0 -3.0 -1.0 -4.0\n",
      "K -1.0  2.0  0.0 -1.0 -3.0  1.0  1.0 -2.0 -1.0 -3.0 -2.0  5.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  0.0  1.0 -1.0 -4.0\n",
      "M -1.0 -1.0 -2.0 -3.0 -1.0  0.0 -2.0 -3.0 -2.0  1.0  2.0 -1.0  5.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0  1.0 -3.0 -1.0 -1.0 -4.0\n",
      "F -2.0 -3.0 -3.0 -3.0 -2.0 -3.0 -3.0 -3.0 -1.0  0.0  0.0 -3.0  0.0  6.0 -4.0 -2.0 -2.0  1.0  3.0 -1.0 -3.0 -3.0 -1.0 -4.0\n",
      "P -1.0 -2.0 -2.0 -1.0 -3.0 -1.0 -1.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -4.0  7.0 -1.0 -1.0 -4.0 -3.0 -2.0 -2.0 -1.0 -2.0 -4.0\n",
      "S  1.0 -1.0  1.0  0.0 -1.0  0.0  0.0  0.0 -1.0 -2.0 -2.0  0.0 -1.0 -2.0 -1.0  4.0  1.0 -3.0 -2.0 -2.0  0.0  0.0  0.0 -4.0\n",
      "T  0.0 -1.0  0.0 -1.0 -1.0 -1.0 -1.0 -2.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  5.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -4.0\n",
      "W -3.0 -3.0 -4.0 -4.0 -2.0 -2.0 -3.0 -2.0 -2.0 -3.0 -2.0 -3.0 -1.0  1.0 -4.0 -3.0 -2.0 11.0  2.0 -3.0 -4.0 -3.0 -2.0 -4.0\n",
      "Y -2.0 -2.0 -2.0 -3.0 -2.0 -1.0 -2.0 -3.0  2.0 -1.0 -1.0 -2.0 -1.0  3.0 -3.0 -2.0 -2.0  2.0  7.0 -1.0 -3.0 -2.0 -1.0 -4.0\n",
      "V  0.0 -3.0 -3.0 -3.0 -1.0 -2.0 -2.0 -3.0 -3.0  3.0  1.0 -2.0  1.0 -1.0 -2.0 -2.0  0.0 -3.0 -1.0  4.0 -3.0 -2.0 -1.0 -4.0\n",
      "B -2.0 -1.0  3.0  4.0 -3.0  0.0  1.0 -1.0  0.0 -3.0 -4.0  0.0 -3.0 -3.0 -2.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
      "Z -1.0  0.0  0.0  1.0 -3.0  3.0  4.0 -2.0  0.0 -3.0 -3.0  1.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
      "X  0.0 -1.0 -1.0 -1.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0  0.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -4.0\n",
      "* -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0  1.0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "print(matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.0\n",
      "MERSTQELFINFTVVLITVLLMWLLVRSYQY\n",
      "|...|.|||.|||.|||||.|||||||||||\n",
      "MGINTRELFLNFTIVLITVILMWLLVRSYQY\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import Align\n",
    "aligner = Align.PairwiseAligner()\n",
    "matrix = Align.substitution_matrices.load(\"BLOSUM62\")\n",
    "aligner.substitution_matrix = matrix\n",
    "aligner.match_score = 1.0\n",
    "aligner.mismatch_score = -2.0\n",
    "aligner.gap_score = -2.5\n",
    "protein_1 = 'MERSTQELFINFTVVLITVLLMWLLVRSYQY'\n",
    "protein_2 = 'MGINTRELFLNFTIVLITVILMWLLVRSYQY'\n",
    "score = aligner.score(protein_1 , protein_2)\n",
    "alignments = aligner.align(protein_1 , protein_2)\n",
    "print(score)\n",
    "for alignment in alignments:\n",
    "    print(alignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Multiple sequence alingmnet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SingleLetterAlphabet() alignment with 6 rows and 26 columns\n",
      "ATCCGTCAGGATAGAGTTTTATTCTC Seq1\n",
      "ATCGGTACGAATAGAGCTTTATTCTC Seq4\n",
      "AACCGTCAGAAGAGAGCTTTATGCAC Seq0\n",
      "AACCGTCCGAATAGCCCTTTATTCTC Seq3\n",
      "AACCGTCCGAATCGAGCATTATTCTC Seq2\n",
      "AACCGTCGGAAAAGAGCTTTAGTCTC Seq5\n"
     ]
    }
   ],
   "source": [
    "#open the alignmnent file\n",
    "from Bio import AlignIO\n",
    "with open(\"aln_file.clustal\", \"r\") as aln:\n",
    "    #use AlignIO to read the alignment file in 'clustal' format\n",
    "    alignment = AlignIO.read(aln, \"clustal\")\n",
    "print(alignment)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 4: Use the alignment to obtain a \"distance\" between all pairs of sequences\n",
    "Using the parsed alignment, we can get the distance (or difference) between all the sequences.\n",
    "\n",
    "This just tells us, for each pair of sequences, how different they are."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Seq1\t0\n",
      "Seq4\t0.1923076923076923\t0\n",
      "Seq0\t0.23076923076923073\t0.2692307692307693\t0\n",
      "Seq3\t0.23076923076923073\t0.1923076923076923\t0.23076923076923073\t0\n",
      "Seq2\t0.23076923076923073\t0.1923076923076923\t0.23076923076923073\t0.15384615384615385\t0\n",
      "Seq5\t0.23076923076923073\t0.23076923076923073\t0.1923076923076923\t0.1923076923076923\t0.1923076923076923\t0\n",
      "\tSeq1\tSeq4\tSeq0\tSeq3\tSeq2\tSeq5\n"
     ]
    }
   ],
   "source": [
    "from Bio.Phylo.TreeConstruction import DistanceCalculator\n",
    "\n",
    "#calculate the distance matrix\n",
    "calculator = DistanceCalculator('identity')\n",
    "#adds distance matrix to the calculator object and returns it\n",
    "dm = calculator.get_distance(alignment)\n",
    "print(dm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Construct a tree from the set of distances\n",
    "And finally, we can construct a phylogenetic tree from the pairwise distances between all the sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tree(rooted=False)\n",
      "    Clade(branch_length=0, name='Inner4')\n",
      "        Clade(branch_length=0.014423076923076886, name='Inner3')\n",
      "            Clade(branch_length=0.024038461538461536, name='Inner1')\n",
      "                Clade(branch_length=0.10096153846153844, name='Seq1')\n",
      "                Clade(branch_length=0.09134615384615385, name='Seq4')\n",
      "            Clade(branch_length=0.024038461538461536, name='Inner2')\n",
      "                Clade(branch_length=0.0801282051282051, name='Seq5')\n",
      "                Clade(branch_length=0.1121794871794872, name='Seq0')\n",
      "        Clade(branch_length=0.07692307692307694, name='Seq3')\n",
      "        Clade(branch_length=0.07692307692307693, name='Seq2')\n"
     ]
    }
   ],
   "source": [
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor\n",
    "\n",
    "#initialize a DistanceTreeConstructor object based on our distance calculator object\n",
    "constructor = DistanceTreeConstructor(calculator)\n",
    "\n",
    "#build the tree\n",
    "tree = constructor.build_tree(alignment)\n",
    "print(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### And let's use the Phylo module to visualize the result!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAEKCAYAAAARnO4WAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzt3X18VOWd9/HPzwhpZSDxAVBEYPQGwTxNSSZKKW3AhVCrtkXTbRVtaYHe4N4vqg2ldpf2xW6zdlv3XrPcVNGG1bZpQ4FtKbaLtrDZUrDmAcdUhRIxsUZAcGuDsYgJXPcfM8wOz5OHk0lyvu/XK6+cOec61/mdw+GXa645c13mnENERAa+C1IdgIiI9A4lfBERn1DCFxHxCSV8ERGfUMIXEfEJJXwREZ9QwhcR8QklfBERn1DCFxHxiQtTHUCiyy67zI0bNy7VYYiI9Bv19fVvOueGJ1O2TyX8cePGUVdXl+owRET6DTN7Ndmy6tIREfEJJXwREZ9QwhcR8QklfBERn1DCFxHxCSV8ERGfUMIXEfEJJXwREZ9QwhcR8QklfBERn1DCFxHxCSV8ERGfUMIXEfEJJXwREZ9QwhcR8QklfBERn1DCFxHxCSV8ERGfUMIXEfEJJXwREZ9QwheRlFuzZg05OTnk5uaSnZ3Nxo0bu1zX7NmzyczM5Oabb+7BCAeGC72q2MzWADcDB51z2V4dR0T6t5aWFsrKyti5cycZGRm0tbVx6NChLte3dOlS/vKXv7B69eoejHJg8LKF/zgw28P6RWQAOHjwIEOHDiUQCAAQCAQIBoMA7N27l9mzZ5Ofn8+0adPYvXs3AE1NTUyZMoVwOMzy5cvj+wLceOONDB06tPdPpB/wLOE7534D/Mmr+kVkYMjLy2PkyJEEg0HmzZvHpk2b4tsWLlzIypUrqa+v58EHH2Tx4sUALFmyhEWLFlFbW8vll1+eqtD7Hc+6dKT3FBUVpToEkU6prq6OL6elpbF582Zqa2vZsmUL9957L/X19ZSWlrJjxw5KSkriZY8ePQrA9u3b2bBhAwB33XUXy5Yt69X4+6uUJ3wzWwgsBBgzZkyKo+l/ioqKiEQihEKhVIci0mVmRmFhIYWFhcycOZN58+Zx3333kZmZSSQSOes+0jkpT/jOuUeBRwEKCgpcisPpl0Kh0EktJpH+ZN++fRw4cIDJkycDEIlEGDt2LMOGDSMYDLJu3TpKSkpwztHQ0EBeXh5Tp06lqqqKuXPnUllZmeIz6D/0WKaIpFR7ezulpaVMnDiRUCjE2rVrKS8vB6CyspKKigry8vLIysqKP65ZXl7OqlWrCIfDtLa2nlTftGnTKCkpYcuWLYwePZqnnnqq18+przLnvGlUm9mPgSLgMuAN4BvOuYpz7VNQUODq6uo8iWegOtF/rxa++FkgEKCtrS3VYaSEmdU75wqSKetZl45z7jNe1S0iIp2nLh0R6ff82rrvLCV8ERGfUMIXEfEJJXwREZ9QwhcR8QklfBERn1DCFxHxCSX8AWjcuHG8+eabnh/n85//PCNGjCA7W9MdiPQHSvjSaceOHQPgc5/7HJs3b05xNCKSLCX8Aay5uZlJkyaxYMECsrKymDVrFkeOHAGiQzIsW7aMwsJCJkyYwLZt24BoMl+6dCnhcJjc3Nz4rEHV1dVMnz6dO+64g5ycHAA+/OEPc8kll6Tm5ESk01I+WmZf19fHmj/f0MiNjY38+Mc/5rHHHuNTn/oUGzZsYO7cuQB0dHRQU1PDL3/5S1asWMGvf/1rKioqyMjIoLa2lqNHjzJ16lRmzZoFQE1NDS+88EJ8NiIR6V+U8M9hIIw1HwwG4/Hn5+fT3Nwc3zZnzpzT1j/99NM0NDSwfv16AFpbW2lsbGTw4MEUFhYq2Yv0Y0r459HXx5o/3zuQ9PT0+HJaWlq8SydxW1paGh0dHQA451i5ciXFxcUn1VNdXc2QIUN6KGoRSQX14ctJiouLefjhh2lvbwdgz549vPPOOymOSkR6ghK+nGT+/Plcd911TJ48mezsbL74xS/GW/+n+sxnPsOUKVP4wx/+wOjRo6moOOd0ByKSYp5NgNIVfW0ClP4wuUh/iFFEvNOZCVDUwhcR8QklfBERn1DCFxHxCSV8ERGfUMIXEfEJJXwREZ9QwhcR8QklfBHpljVr1pCTk0Nubi7Z2dls3LixW/UdPnyYK6+8kr/5m7/poQjlBM/G0jGzq4DvA5cDx4FHnXPlXh1PRHpfS0sLZWVl7Ny5k4yMDNra2jh06FC36ly+fDkf+chHeihCSeRlC78D+LJzbhJwA3CPmV3n4fFEpJcdPHiQoUOHEggEAAgEAvERVffu3cvs2bPJz89n2rRp7N69G4CmpiamTJlCOBxm+fLl8X0B6uvreeONN+JDckvP8izhO+f2O+d2xpbfBnYBV3p1PBHpfXl5eYwcOZJgMMi8efPYtGlTfNvChQtZuXIl9fX1PPjggyxevBiAJUuWsGjRImpra7n88svj5Y8fP86Xv/xlvvOd7/T6efhFrwyPbGbjgA8Az/bG8fwmEon0+YlaZOBIHLcpLS2NzZs3U1tby5YtW7j33nupr6+ntLSUHTt2UFJSEi979OhRALZv386GDRsAuOuuu1i2bBkA3/3ud7npppu46qqreu9kfMbzhG9mAWAD8CXn3OEzbF8ILAQYM2aM1+EMONXV1Ur2klJmRmFhIYWFhcycOZN58+Zx3333kZmZSSQSOes+p3rmmWfYtm0b3/3ud2lra+O9994jEAjwrW99y+tT8A1PE76ZDSKa7Cudc/9+pjLOuUeBRyE6WqaX8QxUGilTUmXfvn0cOHCAyZMnA9F3m2PHjmXYsGEEg0HWrVtHSUkJzjkaGhrIy8tj6tSpVFVVMXfuXCorK+N1JS4//vjj1NXVKdn3MM/68C36J7wC2OWc+79eHUdEUqe9vZ3S0lImTpxIKBRi7dq1lJdHH8arrKykoqKCvLw8srKy4o9rlpeXs2rVKsLhMK2trakM33c8Gw/fzD4EbAN+T/SxTICvOed+ebZ9NB6+iP8EAgHa2tpSHUa/1Znx8D3r0nHO/RY4vaNORERSQt+0FZGUUuu+9yjhi4j4hBK+iIhPKOGLiPiEEr6IiE8o4YuI+IQSfieNGzeON99809NjvPvuuxQWFsa/sPKNb3zD0+OJiD/0yuBpkrxjx46Rnp7O1q1bCQQCtLe386EPfYiPfvSj3HDDDakOT0T6MbXwu6i5uZlJkyaxYMECsrKymDVrFkeOHAGi39BdtmwZhYWFTJgwgW3btgHRZL506VLC4TC5ubmsXr0aiH6Td/r06dxxxx3k5ORgZvExwtvb22lvbz/jYFMiIp2hhN8NjY2N3HPPPbz44otkZmbGh3wF6OjooKamhoceeogVK1YAUFFRQUZGBrW1tdTW1vLYY4/R1NQEQE1NDWVlZbz00ktA9I9DKBRixIgRzJw5k+uvv773T1BEBpR+3aXj9bDAkUiEUCh01u3BYDC+PT8/n+bm5vi2OXPmnLb+6aefpqGhgfXr1wPQ2tpKY2MjgwcPprCwMD5TEETHGY9EIvz5z3/mk5/8JC+88ALZ2dk9fIYi4if9toVfVFR01rG2e0t6enp8OS0tjY6OjtO2Ja53zrFy5UoikQiRSISmpqb4VG5Dhgw54zEyMzMpKipi8+bNXp2GiPhEv27hh0IhT0ey7Ol3EMXFxTz88MPMmDGDQYMGsWfPHq688vRZHw8dOsSgQYPIzMzkyJEj/PrXv47PCiQi0lX9OuH3N/Pnz6e5uZnJkyfjnGP48OH87Gc/O63c/v37+exnP8uxY8c4fvw4n/rUp7j55ptTELGIDCSejYffFZ0ZD783xqrXePgi0td1Zjz8ftuHLyIinaOELyLiE0r4IiI+oYQvIuITSvgiIj6hhC8i4hNK+CIiPqGELyLiE0r4Ij60Zs0acnJyyM3NJTs7m40bN3a5rrS0NEKhEKFQiFtvvbUHo5Se5tnQCmb2PuA3QHrsOOudc5q6SSTFWlpaKCsrY+fOnWRkZNDW1sahQ4e6XN/73//+lA9kKMnxsoV/FJjhnMsDQsBsM9OUTSIpdvDgQYYOHRqfZCcQCMSH5t67dy+zZ88mPz+fadOmsXv3bgCampqYMmUK4XCY5cuXx/eV/sWzhO+i2mIvB8V++s7APSI+lZeXx8iRIwkGg8ybN49NmzbFty1cuJCVK1dSX1/Pgw8+yOLFiwFYsmQJixYtora2lssvv/yk+t59910KCgq44YYbzjgYoPQdno6WaWZpQD3wv4BVzrlnvTyeX3k9EYz0f4kDAKalpbF582Zqa2vZsmUL9957L/X19ZSWlrJjxw5KSkriZY8ePQrA9u3b4zO63XXXXScN1/3HP/6RUaNG8corrzBjxgxycnK45ppreufEpFM8/dDWOXfMORcCRgOFZnbalE1mttDM6sysrjv9iH7VFyaCkf7HzCgsLOT++++nqqqKDRs2cPz4cTIzM+MT9EQiEXbt2nXSPmcyatQoAK6++mqKiop47rnneuUcpPN6ZTx859yfzawamA28cMq2R4FHITo8cm/EM9B4PRGMDCz79u3jwIEDTJ48GYhO5Tl27FiGDRtGMBhk3bp1lJSU4JyjoaGBvLw8pk6dSlVVFXPnzqWysjJe11tvvcVFF11Eeno6b775Jtu3b+crX/lKqk5NzsOzFr6ZDTezzNjy+4G/AnZ7dTwRSU57ezulpaVMnDiRUCjE2rVrKS8vB6CyspKKigry8vLIysqKP65ZXl7OqlWrCIfDtLa2xuvatWsXBQUF5OXlMX36dL761a9y3XXXpeS85Py8bOFfATwR68e/APiJc+5JD48nIkkYO3YsW7duPeO2YDB4xvmTg8EgzzzzTPz1N7/5TQA++MEP8vvf/96bQKXHeZbwnXMNwAe8ql9ERDpH37QVkU5ra2s7fyHpc5TwRUR8IukuHTMbAbzvxGvn3B89iUhERDxx3ha+md1qZo1AE/BfQDPwHx7HJSIiPSyZLp1/AG4A9jjngsCNwHZPoxIRkR6XTMJvd879N3CBmV3gnPtPooOhiYhIP5JMwv+zmQWIDnVcaWblQIe3YUl3jBs3jjfffNPTY7z22mtMnz6dSZMmkZWVFf/ijoj0Xcl8aPtx4F3gXuBOIAP4ey+Dkr7t2LFjXHjhhfzzP/8zkydP5u233yY/P5+ZM2fqW5YifVgyLfyxsUHQOpxzTzjn/hXI8Tow6b7m5mYmTZrEggULyMrKYtasWRw5cgSIDrq2bNkyCgsLmTBhAtu2bQOiyXzp0qWEw2Fyc3NZvXo1EB1tcfr06dxxxx3k5ORwxRVXxMdiGTp0KJMmTeL1119PzYmKSFKSSfg/MbNlFvV+M1sJPOB1YNIzGhsbueeee3jxxRfJzMyMD3EL0NHRQU1NDQ899BArVqwAoKKigoyMDGpra6mtreWxxx6jqakJgJqaGsrKynjppZdOOkZzczPPPfcc119/fe+dmIh0WjJdOtcD/wTsAIYClcBUL4PqSyKRSJ8ebz4SiRAKnf0z9GAwGN+en59Pc3NzfNucOXNOW//000/T0NDA+vXrAWhtbaWxsZHBgwdTWFgYnxnphLa2Nm677TYeeughhg0b1oNnJiI9LZmE3w4cAd5P9ItXTc65455G1UdUV1f36WSfjPT09PhyWlpavEsncVtaWhodHdHP4Z1zrFy5kuLi4pPqqa6uZsiQISeta29v57bbbuPOO++M//EQkb4rmYRfC2wEwsClwGozu905d7unkfURfX2c+Z7+g1RcXMzDDz/MjBkzGDRoEHv27OHKK688rZxzji984QtMmjSJ++67r0djEBFvJJPwv+Ccq4stHwA+bmZ3eRiTpND8+fNpbm5m8uTJOOcYPnz4Gecp3b59Oz/4wQ/IycmJdxn94z/+IzfddFNvhywiSTLnkptkqjfG0ikoKHB1dXXnL8j/tGz7egvca7oOIv5mZvXOuYJkyiYzls4tGktHRKT/S+axzG+isXRERPo9jaUjIuITyXxoe+pYOgeJPqopIiL9SDIJ/3ngL5w8lk7Ay6BERKTnJZPwp8e+aHUceALAzBo8jUpERHrcWRO+mS0CFgPXnJLgh6IPbUVE+p1ztfB/RPTxyweAryasf9s59ydPoxIRkR531qd0nHOtzrlm59xnnHOvJvwo2YtIn7VmzRpycnLIzc0lOzubjRs3drmuJ554gvHjxzN+/HieeOKJHowyNZLpw+8WM0sD6oDXnXM3e308EfGvlpYWysrK2LlzJxkZGbS1tXHo0KEu1fWnP/2JFStWUFdXh5mRn5/PrbfeysUXX9zDUfeeZJ7D764lwK5eOI6I+NzBgwcZOnQogUD0QcJAIBAf0nvv3r3Mnj2b/Px8pk2bxu7duwFoampiypQphMNhli9fHt/3qaeeYubMmVxyySVcfPHFzJw5k82bN6fmxHqIpwnfzEYDHwO+5+VxREQA8vLyGDlyJMFgkHnz5rFp06b4toULF7Jy5Urq6+t58MEHWbx4MQBLlixh0aJF1NbWcvnll8fLv/7661x11VXx16NHj+73s7p53aXzEPAVok/2iEf6+iQtIl5KHDgwLS2NzZs3U1tby5YtW7j33nupr6+ntLSUHTt2UFJSEi979OhRIDry64mZ4O666y6WLVsGRIcAP5WZeXgm3vOshW9mNwMHnXP15ym30MzqzKyuq31tflZdXX3OGa9E/MbMKCws5P7776eqqooNGzZw/PhxMjMziUQi8Z9du3adtM+pRo8ezWuvvRZ/3dLSwqhRo3rlHLziZQt/KnCrmd1EdFjlYWb2Q+fc3MRCzrlHgUchOjyyh/EMWBoaWSRq3759HDhwgMmTJwPRd79jx45l2LBhBINB1q1bR0lJCc45GhoayMvLY+rUqVRVVTF37lwqKyvjdRUXF/O1r32Nt956C4hO//nAA/17Om/PWvjOufudc6Odc+OATwNbT032IiI9qb29ndLSUiZOnEgoFGLt2rWUl5cDUFlZSUVFBXl5eWRlZcUf1ywvL2fVqlWEw2FaW1vjdV1yySUsX76ccDhMOBzm61//OpdccklKzqunJD0BSrcOYlYElJ7vsczuToAybtw46urquOyyy7oYafKOHTtGQUEBV155JU8++aTnxxOR3hEIBGhra0t1GEnr0QlQeoJzrnogPIN/7Nix+HJ5eTmTJk1KYTQiIp3TKwm/tzU3NzNp0iQWLFhAVlYWs2bN4siRI0D0ncGyZcsoLCxkwoQJbNu2DYgm86VLlxIOh8nNzWX16tVA9B3E9OnTueOOO8jJyQGiH9784he/YP78+ak5QRHxTH9q3XfWgEz4AI2Njdxzzz28+OKLZGZmxh+7Aujo6KCmpoaHHnqIFStWAFBRUUFGRga1tbXU1tby2GOP0dTUBEBNTQ1lZWW89NJLAHzpS1/i29/+NhdcMGAvn4gMQJ4PrZAqwWAw/rhifn4+zc3N8W1z5sw5bf3TTz9NQ0MD69evB6C1tZXGxkYGDx5MYWFh/Nt6Tz75JCNGjCA/P19Px4hIvzJgE356enp8OS0tLd6lk7gtLS2Njo4OIPoli5UrV1JcXHxSPdXV1QwZMiT+evv27fz85z/nl7/8Je+++y6HDx9m7ty5/PCHP/TydEREuk19EjHFxcU8/PDDtLdHZ2/cs2cP77zzzmnlHnjgAVpaWmhubqaqqooZM2Yo2YtIvzBgW/idNX/+fJqbm5k8eTLOOYYPH87PfvazVIclItJjeuU5/GR19zl8ERG/6XPP4YuISOop4YuI+IQSvoiITyjhi4j4hBK+iIhPKOGLiPiEEr6IiE8o4YuI+IQSvoiITyjhi4j4hBK+iIhPKOGLiPiEEr6IiE8o4YuI+IQSvoiITyjhi4j4hBK+DFhr1qwhJyeH3NxcsrOz2bhxY5fqefXVV8nPzycUCpGVlcUjjzzSw5GK9A5Ppzg0s2bgbeAY0JHsrCwi3dXS0kJZWRk7d+4kIyODtrY2Dh061KW6rrjiCnbs2EF6ejptbW1kZ2dz6623MmrUqB6OWsRbvdHCn+6cCynZS286ePAgQ4cOJRAIABAIBAgGgwDs3buX2bNnk5+fz7Rp09i9ezcATU1NTJkyhXA4zPLly+P7Dh48mPT0dACOHj3K8ePHU3BGIt3Xrycxj0Qi8bltRRLnN87Ly2PkyJEEg0FuvPFG5syZwy233ALAwoULeeSRRxg/fjzPPvssixcvZuvWrSxZsoRFixZx9913s2rVqpPqfu211/jYxz7Gyy+/zHe+8x217qVf8nQSczNrAt4CHLDaOffoGcosBBYCjBkzJv/VV19Nqu6ioiIikQihUKgHI5b+7NQJ7Z1z1NbWsmXLFioqKpg7dy6lpaUMHz6ca6+9Nl7u6NGj7Nq1i0svvZQDBw4waNAgDh8+zKhRo2hrazupzn379vGJT3yCTZs2MXLkyN44LZFz6swk5l638Kc65/aZ2QjgV2a22zn3m8QCsT8CjwIUFBR06q9PKBQ67T+5yAlmRmFhIYWFhcycOZN58+Zx3333kZmZSSQSOes+5zJq1CiysrLYtm0bt99+uxdhi3jG0z5859y+2O+DwE+BQi+PJ3LCvn372LlzZ/x1JBJh7NixDBs2jGAwyLp164Dou4Dnn38egKlTp1JVVQVAZWVlfN+WlhaOHDkCwFtvvcX27dtPeocg0l94lvDNbIiZDT2xDMwCXvDqeCKJ2tvbKS0tZeLEiYRCIdauXUt5eTkQTeYVFRXk5eWRlZUVf1yzvLycVatWEQ6HaW1tjde1a9curr/+evLy8vjIRz5CaWkpOTk5KTkvke7wrA/fzK4m2qqHaNfRj5xzZefap6CgwNXV1SVV/4kPa9WlI14JBAKn9eGL9DV9og/fOfcKkOdV/SIi0jn6pq3IWah1LwONEr6IiE8o4YuI+IQSvoiITyjhi4j4hBK+iIhPKOGLiPiEEr6IiE8o4YuI+IQSvoiITyjhi4j4hBK+iIhPKOGLiPiEEr6IiE8o4YuI+IQSvoiITyjhi4j4hBK+iIhPKOGLiPiEEr6IiE8o4YuI+IQSvoiITyjhi4j4hBK+DFhr1qwhJyeH3NxcsrOz2bhxY5fqiUQiTJkyhaysLHJzc1m7dm0PRyrSOy70snIzywS+B2QDDvi8c+4ZL48pAtDS0kJZWRk7d+4kIyODtrY2Dh061KW6LrroIr7//e8zfvx49u3bR35+PsXFxWRmZvZw1CLe8rqFXw5sds5NBPKAXR4fTwSAgwcPMnToUAKBAACBQIBgMAjA3r17mT17Nvn5+UybNo3du3cD0NTUxJQpUwiHwyxfvjy+74QJExg/fjwAo0aNYsSIEV3+4yGSSp618M1sGPBh4HMAzrn3gPd68hiRSISioqKerFL6serq6vhyXl4eI0eOJBgMcuONNzJnzhxuueUWABYuXMgjjzzC+PHjefbZZ1m8eDFbt25lyZIlLFq0iLvvvptVq1ad8Rg1NTW89957XHPNNb1xSiI9ypxz3lRsFgIeBV4i2rqvB5Y45945pdxCYCHAmDFj8l999dWkj6FkL4kSEz6Ac47a2lq2bNlCRUUFc+fOpbS0lOHDh3PttdfGyx09epRdu3Zx6aWXcuDAAQYNGsThw4cZNWoUbW1t8XL79++nqKiIJ554ghtuuKG3TkvknMys3jlXkFRZDxN+AfA7YKpz7lkzKwcOO+eWn22fgoICV1dX50k84m91dXXMmzeP7du3c+2117J///7Tylx66aW88cYbXHjhhacl/MOHD1NUVMT9999PSUlJb4cvcladSfhe9uG3AC3OuWdjr9cDkz08nkjcvn372LlzZ/x1JBJh7NixDBs2jGAwyLp164Dou4Dnn38egKlTp1JVVQVAZWVlfN/33nuPT37yk9x9991K9tKveZbwnXMHgNfM7MR75xuJdu+IeK69vZ3S0lImTpxIKBRi7dq1lJeXA9FkXlFRQV5eHllZWfHHNcvLy1m1ahXhcJjW1tZ4XT/5yU/4zW9+w+OPP04oFCIUChGJRFJyXiLd4VmXDsT78b8HDAZeAeY55946W3l16UhfEggETurDF+mLOtOl4+lz+M65CJBUICIi4i1901bkLNS6l4FGCV9ExCeU8EVEfEIJX0TEJ5TwRUR8QglfRMQnlPBFRHxCCV9ExCeU8EVEfEIJX0TEJ5TwRUR8QglfRMQnlPBFRHxCCV9ExCeU8EVEfEIJX0TEJ5TwRUR8QglfRMQnlPBFRHxCCV9ExCeU8EVEfMKcc6mOIc7MDgGvdmKXy4A3PQqnpylWb/SXWPtLnKBYveJVrGOdc8OTKdinEn5nmVmdc64g1XEkQ7F6o7/E2l/iBMXqlb4Qq7p0RER8QglfRMQn+nvCfzTVAXSCYvVGf4m1v8QJitUrKY+1X/fhi4hI8vp7C19ERJLUpxK+mc02sz+Y2ctm9tUzbE83s7Wx7c+a2biEbffH1v/BzIqTrbO3YzWzmWZWb2a/j/2ekbBPdazOSOxnRArjHGdmRxJieSRhn/xY/C+b2b+amXU3zm7GemdCnBEzO25modi2Hr+mScb6YTPbaWYdZnb7Kds+a2aNsZ/PJqzv8eva1TjNLGRmz5jZi2bWYGZ/nbDtcTNrSrimoe7G2Z1YY9uOJcTz84T1wdi90hi7dwanMlYzm37KvfqumX0its2T63oS51yf+AHSgL3A1cBg4HngulPKLAYeiS1/GlgbW74uVj4dCMbqSUumzhTE+gFgVGw5G3g9YZ9qoKCPXNNxwAtnqbcGmAIY8B/AR1MZ6yllcoBXvLqmnYh1HJALfB+4PWH9JcArsd8Xx5Yv9uK6djPOCcD42PIoYD+QGXv9eGLZVF/T2La2s9T7E+DTseVHgEWpjvWUe+FPwEVeXddTf/pSC78QeNk594pz7j2gCvj4KWU+DjwRW14P3BhrBX0cqHLOHXXONQEvx+pLps5ejdU595xzbl9s/YvA+8wsvQdi6tE4z1ahmV0BDHPOPeOid+n3gU/0oVg/A/y4B+I5l/PG6pxrds41AMdP2bcY+JVz7k/OubeAXwGzPbquXY7TObfHOdcYW96nAMIFAAAGVElEQVQHHASS+nJPb8d6NrF7YwbRewWi906v3KtJxno78B/Oub/0QExJ6UsJ/0rgtYTXLbF1ZyzjnOsAWoFLz7FvMnX2dqyJbgOec84dTVj3b7G3c8t74C19d+MMmtlzZvZfZjYtoXzLeepMRawn/DWnJ/yevKbJxtrZfb24rj1y/5tZIdGW7N6E1WWxrp5/6aEGS3djfZ+Z1ZnZ7050kRC9N/4cu1e6UufZ9FRe+TSn36s9fV1P0pcS/pn+I576CNHZynR2fXd1J9boRrMs4J+ALyZsv9M5lwNMi/3clcI49wNjnHMfAO4DfmRmw5Kssyt64ppeD/zFOfdCwvaevqbnjaOL+3pxXbtdZ+ydxw+Aec65E63V+4GJQJhot8Sy7gR54lBnWNeZWMe46LdY7wAeMrNreqDOs+mp65oDPJWw2ovrepK+lPBbgKsSXo8G9p2tjJldCGQQ7QM7277J1NnbsWJmo4GfAnc75+KtJufc67HfbwM/IvrWMSVxxrrH/jsWTz3R1t2EWPnR56mzV2NN2H5ai8mDa5psrJ3d14vr2q37P/YH/hfA3znnfndivXNuv4s6Cvwbqb+mJ7qdcM69QvRzmw8QHbcmM3avdLpOr2KN+RTwU+dc+4kVHl3Xk/SlhF8LjI99qj6Y6H/en59S5ufAiacabge2xvo7fw582qJPcQSB8UQ/AEumzl6N1cwyif4nut85t/1EYTO70Mwuiy0PAm4GXqB7uhPncDNLi8VzNdFr+opzbj/wtpndEOseuRvY2M04uxVrLMYLgBKi/anE1nlxTZON9WyeAmaZ2cVmdjEwC3jKo+va5Thj5X8KfN85t+6UbVfEfhvRPvGUXtPYtUyPLV8GTAVeit0b/0n0XoHovdNb9+r5nPZZk0fX9WRefiLc2R/gJmAP0dbk38bW/T1wa2z5fcA6oh/K1gBXJ+z7t7H9/kDC0w1nqjOVsQJ/B7wDRBJ+RgBDgHqggeiHueVAWgrjvC0Wx/PATuCWhDoLiN6Me4H/R+wLfCn+9y8CfndKfZ5c0yRjDRNtCb4D/DfwYsK+n4+dw8tEu0o8u65djROYC7Sfcp+GYtu2Ar+PxfpDIJDKawp8MBbP87HfX0io8+rYvfJy7N5J7wP//uOA14ELTqnTk+ua+KNv2oqI+ERf6tIREREPKeGLiPiEEr6IiE8o4YuI+IQSvoiITyjhS59l0RE7e/5Z5HMfs9rMzjnvqFdxmVmRmX0w4fXjdsqokCLdoYQv/d6JL4gNAEVEnykX8YQSvvR1F5rZE7EBpdab2UUAZtZsZl83s98CJWa2wMxqzex5M9uQUO5xi44tv8PMXrGTxyb/ikXHn3/ezL6VcMwSM6sxsz0Jg8adkZmlmdl3YsduMLMvxtYXxd4trDez3WZWGfsGJWZ2U2zdb2OxPWnRsf3/N3CvRQd6O3HcD58pdpGuUMKXvu5a4FHnXC5wmOiY+Ce865z7kHOuCvh351zYOZcH7AK+kFDuCuBDRIdW+BaAmX2U6NfXr4/t8+2E8hc65wqBLwHfOE98XwBanXNhot+uXBAb3gOi47l8ieh8DVcDU83sfcBqot8G/xCxIYedc81Ex2v/F+dcyDm37Wyxi3SVEr70da+5/xlz6IdEk98JaxOWs81sm5n9HrgTyErY9jPn3HHn3EvAyNi6vwL+zcXGInfOJQ7C9u+x3/VEvwZ/LrOAu80sAjxLdEje8bFtNc65FhcdZTISq2si0TGJmmJlzjd2/5liF+mSC89fRCSlTh37I/H1OwnLjwOfcM49b2afI9offkLifAOW8Pts44qcKH+M8/8fMeD/OOeeOmmlWdEpxz1RV2fH4z9T7CJdoha+9HVjzGxKbPkzwG/PUm4osD82KuadSdT7NPD5hL7+S7oY31PAothxMbMJZjbkHOV3A1fb/8zH/NcJ294meh4inlDCl75uF/BZM2sgOinEw2cpt5xol8qviCbVc3LObSY6pG1drDumtIvxfQ94CdgZe1RzNed4V+CcO0L0c4jNsQ+c3yA6cxfAJuCTp3xoK9JjNFqmSC8zs4Bzri321M4qoNE59y+pjksGPrXwRXrfgti7iheJztq1OsXxiE+ohS8i4hNq4YuI+IQSvoiITyjhi4j4hBK+iIhPKOGLiPiEEr6IiE/8f/+mKvgQIjuAAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "from Bio import Phylo\n",
    "import pylab\n",
    "#draw the tree\n",
    "Phylo.draw(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NCBI BLAST "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Alphabet import generic_dna\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Seq import Seq\n",
    "sequence = Seq(\"ATTTTCTTGCTCTTGAGCTCTGGCACTTCTCTGCTGCTGTC\", generic_dna)\n",
    "result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", sequence)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****Alignment****\n",
      "sequence: gi|31790676|dbj|AP002353.4| Homo sapiens genomic DNA, chromosome 11 clone:CTD-2651C21, complete sequence\n",
      "length: 202607\n",
      "e value: 5.15868e-11\n",
      "ATTTTCTTGCTCTTGAGCTCTGGCACTTCTCTGCTGCTGTC...\n",
      "|||||||||||||||||||||||||||||||||||||||||...\n",
      "ATTTTCTTGCTCTTGAGCTCTGGCACTTCTCTGCTGCTGTC...\n",
      "****Alignment****\n",
      "sequence: gi|694946907|ref|XR_673052.1| PREDICTED: Pan troglodytes sarcolipin (SLN), transcript variant X2, misc_RNA\n",
      "length: 904\n",
      "e value: 2.67226e-08\n",
      "CTTGCTCTTGAGCTCTGGCACTTCTCTGCTGCTGTC...\n",
      "||||||||||||||||||||||||||||||||||||...\n",
      "CTTGCTCTTGAGCTCTGGCACTTCTCTGCTGCTGTC...\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "blast_records = NCBIXML.parse(result_handle)\n",
    "for b in blast_records:\n",
    "    for alignment in b.alignments[:2]:\n",
    "        for hsp in alignment.hsps:\n",
    "            print('****Alignment****')\n",
    "            print('sequence:', alignment.title)\n",
    "            print('length:', alignment.length)\n",
    "            print('e value:', hsp.expect)\n",
    "            print(hsp.query[0:75] + '...')\n",
    "            print(hsp.match[0:75] + '...')\n",
    "            print(hsp.sbjct[0:75] + '...')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Handling PDB files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading PDB structure '1mbs'...\n",
      "Desired structure doesn't exists\n",
      "Downloading PDB structure '1mbs'...\n",
      "Desired structure doesn't exists\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'.\\\\1mbs.cif'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.PDB import PDBList\n",
    "pdbl = PDBList() \n",
    "pdbl.retrieve_pdb_file('1mbs', pdir = '.', file_format = 'pdb')\n",
    "pdbl = PDBList() \n",
    "pdbl.retrieve_pdb_file('1mbs', pdir = '.', file_format = 'mmCif')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['name', 'head', 'idcode', 'deposition_date', 'release_date', 'structure_method', 'resolution', 'structure_reference', 'journal_reference', 'author', 'compound', 'source', 'has_missing_residues', 'missing_residues', 'keywords', 'journal'])\n",
      "-----------------------------\n",
      " x-ray crystallographic studies of seal myoglobin. the molecule at 2.5 angstroms resolution\n",
      "1979-05-15\n",
      "2.5\n",
      "oxygen transport\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB.PDBParser import PDBParser\n",
    "\n",
    "parser = PDBParser(QUIET = True)\n",
    "pdb_id = \"1mbs\"\n",
    "pdb_filename = \"pdb1mbs.ent\"\n",
    "\n",
    "structure_pdb = parser.get_structure(pdb_id, pdb_filename)\n",
    "print(structure_pdb.header.keys())\n",
    "print('-----------------------------')\n",
    "print(structure_pdb.header[\"name\"]) \n",
    "print(structure_pdb.header[\"release_date\"]) \n",
    "print(structure_pdb.header[\"resolution\"])\n",
    "print(structure_pdb.header[\"keywords\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['52.12']\n",
      "1266\n",
      "['11.178', '10.462', '9.483', '8.979', '9.240', '8.164', '7.481', '7.424', '7.101', '7.724']\n"
     ]
    }
   ],
   "source": [
    "print(mmcif_dict[\"_exptl_crystal.density_percent_sol\"])\n",
    "y_list = mmcif_dict[\"_atom_site.Cartn_y\"] # All y coordinates\n",
    "print(len(y_list))\n",
    "print(y_list[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[<Model id=0>]\n"
     ]
    }
   ],
   "source": [
    "models = list(structure_pdb.get_models())\n",
    "print(models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[<Chain id=A>]\n"
     ]
    }
   ],
   "source": [
    "chains = list(models[0].get_chains()) \n",
    "print(chains)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "154\n"
     ]
    }
   ],
   "source": [
    "residues_0 = list(chains[0].get_residues())\n",
    "print(len(residues_0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Residue GLY het=  resseq=1 icode= >\n",
      "[<Atom N>, <Atom CA>, <Atom C>, <Atom O>]\n",
      "<Vector 15.74, 11.18, -11.73>\n"
     ]
    }
   ],
   "source": [
    "print(residues_0[0])\n",
    "atoms = list(residues_0[0].get_atoms())\n",
    "print(atoms)\n",
    "print(atoms[0].get_vector())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Atom N>\n",
      "<Vector 15.74, 11.18, -11.73>\n",
      "<Atom CA>\n",
      "<Vector 15.23, 10.46, -10.56>\n",
      "<Atom C>\n",
      "<Vector 16.28, 9.48, -10.00>\n",
      "<Atom O>\n",
      "<Vector 17.15, 8.98, -10.71>\n"
     ]
    }
   ],
   "source": [
    "for model in structure_pdb.get_models():\n",
    "    for chain in model.get_chains():        \n",
    "        for residue in chain.get_residues():\n",
    "            for atom in residue.get_atoms():\n",
    "                print(atom)\n",
    "                print(atom.get_vector())\n",
    "            break\n",
    "        break\n",
    "    break\n",
    "            \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['data_', '_entry.id', '_audit_conform.dict_name', '_audit_conform.dict_version', '_audit_conform.dict_location', '_database_2.database_id', '_database_2.database_code', '_pdbx_database_status.status_code', '_pdbx_database_status.entry_id', '_pdbx_database_status.recvd_initial_deposition_date']\n",
      "528\n"
     ]
    }
   ],
   "source": [
    "from Bio.PDB.MMCIF2Dict import MMCIF2Dict\n",
    "mmcif_dict = MMCIF2Dict(\"1mbs.cif\")\n",
    "print(list(mmcif_dict.keys())[:10])\n",
    "print(len(list(mmcif_dict.keys())))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Distance between two atoms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.805758\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "coordinates_0 = atoms[0].get_vector()\n",
    "coordinates_1 = atoms[-1].get_vector()\n",
    "x0,y0,z0 = coordinates_0\n",
    "x1,y1,z1 = coordinates_1\n",
    "distance = math.sqrt((x1 - x0)**2 + (y1 - y0)**2 + (z1 - z0)**2)\n",
    "print(round(distance,6))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.8057582\n"
     ]
    }
   ],
   "source": [
    "atoms = list(residues_0[0].get_atoms())\n",
    "print(atoms[0]-atoms[-1])#builtin in pdb module"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise\n",
    "<ol>\n",
    "<li>Create a DNA &lsquo;seq&rsquo; object using sequence. &lsquo;ATCCGGAATCCG&rsquo;.</li>\n",
    "<li>With the above seq object, find its reverse complement and its translated protein sequence.</li>\n",
    "<li>Read the fasta file &lsquo;exercise.fasta&rsquo; provided in the supplementary material and print the sequences.</li>\n",
    "<li>The above sequences are 16s mRNAs of certain microorganisms. Perform BLAST on the identify the species.</li>\n",
    "<li>Apply multiple sequence alignment on the sequences and construct a phylogenetic tree using BioPython Packages.</li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
